core/cc/socket_connection.cpp

```diff
@@ -258,20 +258,81 @@ std::unique_ptr<Connection> SocketConnection::createSocket(
     printf("Bound on port '%d'\n", ntohs(sin.sin_port));
     fflush(stdout); // Force the message for piped readers
 
-    const char* portFile = getenv("GAPII_PORT_FILE");
-    if (portFile != nullptr) {
-        FILE* f = fopen(portFile, "w");
-        if (f != nullptr) {
-            fprintf(f, "Bound on port '%d'", ntohs(sin.sin_port));
-            fclose(f);
-        }
-    }
+  writePortToFile(ntohs(sin.sin_port));
 
     sockScopeGuard.release();
     return std::unique_ptr<Connection>(new SocketConnection(sock));
 }
 
-std::unique_ptr<Connection> SocketConnection::createPipe(const char* pipename, bool abstract) {
+bool SocketConnection::writePortToFile(uint32_t port) {
+  const char* portFile = getenv("GAPII_PORT_FILE");
+  if (portFile == nullptr) {
+    return false;
+  }
+  FILE* f = fopen(portFile, "w");
+  if (f == nullptr) {
+    return false;
+  }
+  fprintf(f, "Bound on port '%d'", port);
+  fclose(f);
+  return true;
+}
+
+uint32_t SocketConnection::getFreePort(const char* hostname) {
```

Comment:
This is just a copy of the createSocket() function, except it just returns the free port assigned by OS, the socket will be closed.